iavl/iavl_tree.go

```diff
@@ -1,37 +1,131 @@
+/*
+	Package merkle implements a key/value store with an immutable Merkle tree.
+*/
+
 package iavl
 
 import (
 	"bytes"
 	"container/list"
+	"fmt"
 	"sync"
 
 	wire "github.com/tendermint/go-wire"
-	. "github.com/tendermint/tmlibs/common"
-	dbm "github.com/tendermint/tmlibs/db"
+	cmn "github.com/tendermint/tmlibs/common"
+	"github.com/tendermint/tmlibs/db"
 	"github.com/tendermint/tmlibs/merkle"
 )
 
+// Keys for administrative persistent data
+var rootsKey = []byte("go-merkle.roots")      // Database key for the list of roots
+var versionKey = []byte("go-merkle.version")  // Database key for the version
+var orphansKey = []byte("go-merkle.orphans/") // Partial database key for each set of orphans
+var deletesKey = []byte("go-merkle.deletes")  // Database key for roots to be pruned
+
+// Fixed number of versions
+const rootsMax = 3
+
+// TODO: Debugging, remove
+var lastId = 0
+
 /*
 Immutable AVL Tree (wraps the Node root)
 This tree is not goroutine safe.
 */
 type IAVLTree struct {
-	root *IAVLNode
-	ndb  *nodeDB
+	ndb     *nodeDB
+	version int
+	roots   *list.List
+	id      int
 }
 
-// NewIAVLTree creates both im-memory and persistent instances
-func NewIAVLTree(cacheSize int, db dbm.DB) *IAVLTree {
+// NewIAVLTree constructor for both persistent and memory based trees
+func NewIAVLTree(cacheSize int, db db.DB) *IAVLTree {
+	//fmt.Printf("NewIAVLTree\n")
+	lastId++
 	if db == nil {
 		// In-memory IAVLTree
-		return &IAVLTree{}
+		return &IAVLTree{
+			version: 0,
+			roots:   list.New(),
+			id:      lastId,
+		}
 	} else {
 		// Persistent IAVLTree
 		ndb := newNodeDB(cacheSize, db)
 		return &IAVLTree{
-			ndb: ndb,
+			ndb:     ndb,
+			version: 0,
+			roots:   list.New(),
+			id:      lastId,
+		}
+	}
+}
+
+// GetValue is a slow way to get the ith element's value
+func GetValue(l *list.List, index int) interface{} {
```

Comment:
Great idea. The names bug me too, but I can live with that for a while...

Is it possible to just attach both to List? Also, we now use List in two places in iavl, and I think the original code (cache vs. cacheQueue) uses a map to speed up random lookups. Maybe utils or common should have a cleaner List implementation that adds the maps and maybe gives some of the primitives more obvious names (like PushFront).